cmd/bosun/expr/funcs.go

```diff
@@ -384,6 +384,161 @@ var builtins = map[string]parse.Func{
 		F:       V,
 		MapFunc: true,
 	},
+	"aggregate": {
+		Args:   []models.FuncType{models.TypeSeriesSet, models.TypeString, models.TypeString},
+		Return: models.TypeSeriesSet,
+		Tags:   tagFirst,
+		F:      Aggregate,
+	},
+}
+
+// Aggregate combines multiple series matching the specified groups using an aggregator function. If group
+// is empty, all given series are combined, regardless of existing groups.
+// Available aggregator functions include: avg (average), p50 (median), min (minimum) and max (maximum).
+func Aggregate(e *State, T miniprofiler.Timer, series *Results, groups string, aggregator string) (*Results, error) {
+	results := Results{}
+
+	grps := splitGroups(groups)
+	if len(grps) == 0 {
+		// no groups specified, so we merge all group values
+		res, err := aggregate(e, T, series, aggregator)
+		if err != nil {
+			return &results, err
+		}
+		res.Group = opentsdb.TagSet{}
+		results.Results = append(results.Results, res)
+	} else {
+		// at least one group specified, so we work out what
+		// the new group values will be
+		newGroups := map[string]*Results{}
+		for _, result := range series.Results {
+			vals := []string{}
+			for _, grp := range grps {
+				if val, ok := result.Group[grp]; ok {
+					vals = append(vals, val)
+				} else {
+					return nil, fmt.Errorf("unmatched group in at least one series: %v", grp)
+				}
+			}
+			groupName := strings.Join(vals, ",")
+			if _, ok := newGroups[groupName]; !ok {
+				newGroups[groupName] = &Results{}
+			}
+			newGroups[groupName].Results = append(newGroups[groupName].Results, result)
+		}
+
+		for groupName, series := range newGroups {
+			res, err := aggregate(e, T, series, aggregator)
+			if err != nil {
+				return &results, err
+			}
+			vs := strings.Split(groupName, ",")
+			res.Group = opentsdb.TagSet{}
+			for i := 0; i < len(grps); i++ {
+				res.Group.Merge(opentsdb.TagSet{grps[i]: vs[i]})
+			}
+			results.Results = append(results.Results, res)
+		}
+	}
+
+	return &results, nil
+}
+
+// Splits a string of groups by comma, but also trims any added whitespace
+// and returns an empty slice if the string is empty.
+func splitGroups(groups string) []string {
+	if len(groups) == 0 {
+		return []string{}
+	}
+	grps := strings.Split(groups, ",")
+	for i, grp := range grps {
+		grps[i] = strings.Trim(grp, " ")
+	}
+	return grps
+}
+
+func aggregate(e *State, T miniprofiler.Timer, series *Results, aggregator string) (*Result, error) {
+	res := Result{}
+	newSeries := make(Series)
+
+	switch aggregator {
```

Comment:
Probably useful to mimic what we do in http://bosun.org/expressions#windowquery-string-duration-string-period-string-num-scalar-funcname-string-seriesset : 

> In addition to supporting Bosun’s reduction functions that take on argument, percentile operations may be be done by setting funcName to p followed by number that is between 0 and 1 (inclusively). For example, "p.25" will be the 25th percentile, "p.999" will be the 99.9th percentile. "p0" and "p1" are min and max respectively (However, in these cases it is recommended to use "min" and "max" for the sake of clarity.

Reply:
Nice, I wasn't aware that it's done that way in the window query. Will make it so 👍